### PR TITLE
docs: Add app / code review tool attestations

### PR DESCRIPTION
**Pull request by adityasaky**
Adds app / code review tool attestations as an extension doc after some discussions with @wlynch re https://github.com/gittuf/github-app.

### CODE REVIEW

File: docs/extensions/code-review-tool-attestations.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+gittuf policy must be updated to extend the information tracked for each trusted
+person. Currently, each trusted person is represented directly by their public
+key in policy metadata, similar to how TUF represents actors. To support code
+review tool attestations, this must be extended to support the identity of an
+approver from the perspective of the code review tool. For example, Alice may
+use an SSH key for commits and RSL entries but use GitHub's cloud offering for
+code review. Thus, gittuf policy must associate her public key and her GitHub
+username / identifier. Also see: https://github.com/gittuf/gittuf/issues/586.
+
+This extension proposes the following format for the policy schema:
```
[the remark this answers is not in the document]
Comment (adityasaky) [reply docs/extensions/code-review-tool-attestations.md:41]: I think I get what you're getting at, but could you rephrase to be certain?

[the remark this answers is not in the document]
Comment (adityasaky) [reply docs/extensions/code-review-tool-attestations.md:41]: Good points, and I'll raise you one!

Right now, I'm focusing on allowing "alice" to sign using any of her keys. I think we could explore parametrizing it in future, like `alice:<keyid>` or something, but I'm unsure how frequently this is going to be necessary. I also see no reason we don't continue supporting `Key` as a type where this is a concern. But overall, I think the point you raise is an open question: do we support setting constraints on a person as to what keys they ought to use in a particular context? What about thresholds? Right now, it only counts towards the threshold for the rule but perhaps in a context, Alice can only count as having signed if she uses _two of her keys_ rather than just one.

The other thing I was considering: consistency in Person definitions. Is a `Person` redefined in multiple rule files where they're delegated to? How do we keep them consistent across definitions?

[the remark this answers is not in the document]
Comment (adityasaky) [reply docs/extensions/code-review-tool-attestations.md:41]: This is really a two-parter:
a) Storing metadata alongside a person definition pertaining to their github etc identities
b) Using those identities to verify code review attestations from SCPs

File: docs/extensions/code-review-tool-attestations.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+## Attestation Schema
+
+The attestation is similar to the reference authorization, but also includes
+information indicating the approvers.
+
+```
+TargetRef    string
+FromTargetID string
+ToTargetID   string
+Approvers    []string
```
Comment (wlynch) [inline docs/extensions/code-review-tool-attestations.md:22]: Worth adding comments + examples for each.

File: docs/extensions/code-review-tool-attestations.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+```
+
+The attestation must indicate the code review tool via the predicate type, and
```
Comment (wlynch) [inline docs/extensions/code-review-tool-attestations.md:25]: > via the predicate type

Where should predicate type be specified?

File: docs/extensions/code-review-tool-attestations.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+review tool attestations, this must be extended to support the identity of an
+approver from the perspective of the code review tool. For example, Alice may
+use an SSH key for commits and RSL entries but use GitHub's cloud offering for
+code review. Thus, gittuf policy must associate her public key and her GitHub
+username / identifier. Also see: https://github.com/gittuf/gittuf/issues/586.
+
+This extension proposes the following format for the policy schema:
+
+```
+"<id>": {
```
Comment (wlynch) [inline docs/extensions/code-review-tool-attestations.md:44]: what would be the primarily id here?

File: docs/extensions/code-review-tool-attestations.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+            "keytype": "sigstore-oidc",
+            "keyval": {
+                "identity": "<username>",
+                "issuer": "<issuer>"
+            },
+            "scheme": "fulcio"
+          },
+        "<keyid2>>": ...
+    },
+    "associatedAppIdentities": {
```
Comment (wlynch) [inline docs/extensions/code-review-tool-attestations.md:59]: bikeshed:

```suggestion
    "associatedIdentities": {
```

I don't think we need to over-index on "app" terminology here. In my mind, technically a human should also be able to make this kind of assertion if they wanted.

File: docs/extensions/code-review-tool-attestations.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        "<appID1>": "<usernameA>",
```
Comment (wlynch) [inline docs/extensions/code-review-tool-attestations.md:60]: I know we also talked about there would be information needed to describe the appID's trust - e.g. the keys it uses, etc. what do we need and where would it be specified?

File: docs/extensions/code-review-tool-attestations.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+during verification, but this has usability concerns. To balance usability and
+security, this extension proposes the use of special attestations created by
+popular code review tools.
+
+## Attestation Schema
+
+The attestation is similar to the reference authorization, but also includes
+information indicating the approvers.
+
+```
```
Comment (wlynch) [inline docs/extensions/code-review-tool-attestations.md:18]: Would the diff codeblock type be useful here to show what's new?

File: docs/extensions/code-review-tool-attestations.md (hunk 3 above)
Comment (adityasaky) [reply docs/extensions/code-review-tool-attestations.md:25]: That's in the attestation, it's a required field in the in-toto statement.

File: docs/extensions/code-review-tool-attestations.md (hunk 4 above)
Comment (adityasaky) [reply docs/extensions/code-review-tool-attestations.md:44]: I wanted to leave it up to the users, it could be anything they want? Right now we use keyids but I don't see why we can't have it be usernames or other identifiers based on the context.

File: docs/extensions/code-review-tool-attestations.md (hunk 6 above)
Comment (adityasaky) [reply docs/extensions/code-review-tool-attestations.md:60]: I think it's just the signer, since all we really have is a signed attestation. The public key is in the root.json file. I think the app ID could be any human picked name again? The root.json would have `appID: <principal>` mappings effectively. `principal` is what I'm calling the interface right now that could be a person, a single pubkey, etc.

File: docs/extensions/code-review-tool-attestations.md (hunk 7 above)
Comment (adityasaky) [reply docs/extensions/code-review-tool-attestations.md:18]: This is a new predicate type altogether rather than an extension of the reference authorization type. I think that makes it easier to ensure we have no confusion in how these attestations are used.

**Comment by adityasaky** [comment]
#626 is a first pass at supporting this in the policy metadata, we could use that to discuss the changes proposed here.

File: docs/gaps/5/README.md [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+        "<integrationID2>": "<usernameB>",
+        ...
+    }
+    "customMetadata": {...}
+}
+```
+
+Here, `personID` can be set to any unique (within the policy) identifier of a
+person such as their name or email address, similar to
+[TAP-12](https://github.com/theupdateframework/taps/blob/master/tap12.md).  Each
```
Comment (patzielinski) [inline docs/gaps/5/README.md:76]:
```suggestion
[TAP-12](https://github.com/theupdateframework/taps/blob/master/tap12.md). Each
```

File: docs/gaps/5/README.md [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+        },
+    },
+    "associatedIdentities": {
+        "https://github.com": "adityasaky+8928778",
+    }
+    "customMetadata": {}
+}
+```
+
+Each associated identity SHOULD be recorded with an immutable value. For
```
Comment (patzielinski) [inline docs/gaps/5/README.md:115]: Do we follow any special meaning for capitalized words like SHOULD, e.g. https://www.rfc-editor.org/rfc/rfc2119?

File: docs/gaps/5/README.md [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+single key or a more complex entity that contains more than one signing key and
+other custom metadata. A principal has the following attributes:
+* ID: A unique identifier for the principal
+* Keys: A set of public keys that can be used to verify signatures, no
+  duplicates are allowed
+* CustomMetadata: Additional custom metadata that must be associated with the
+  principal
+
+Additionally, this GAP introduces the first new principal type, to represent a
+"person".  A "person" has the following schema:
```
Comment (patzielinski) [inline docs/gaps/5/README.md:47]:
```suggestion
"person". A "person" has the following schema:
```

File: docs/gaps/5/README.md [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+existing policies that use keys directly.
+
+## Security
+
+This GAP introduces some concepts that have security implications.
+
+### Out of Sync Principal Definitions
+
+The delegating entity is responsible for associating a unique identifier for the
+delegatee "person".  While this is unique within the context of a delegated
```
Comment (patzielinski) [inline docs/gaps/5/README.md:184]:
```suggestion
delegatee "person". While this is unique within the context of a delegated
```

File: docs/gaps/5/README.md [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+file, other delegated files may use the same ID for another person. To avoid
+policy misconfiguration or confusion issues, the person definitions SHOULD be
+consistent across metadata.  In other words, the definition for `Bob` SHOULD be
```
Comment (patzielinski) [inline docs/gaps/5/README.md:187]:
```suggestion
consistent across metadata. In other words, the definition for `Bob` SHOULD be
```

**Review by patzielinski — APPROVED** [review]
I think we can merge this and iterate on it with updates as needed.

File: docs/gaps/6/README.md [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+Then, the attestations's `Approvers` are used to extend the list of `personIDs`
+verified via their signatures. Each approver is matched to a person using gittuf
+policy, and the corresponding `personID` is added to `personIDs`. Note that
+`personIDs` must contain no duplicates, to ensure the same person is not counted
+twice towards the threshold. If the threshold is met with the approvers in the
+code review tool attestation, then verification passes.
+
+## Reasoning
+
+TODO: flesh out reasoning.
```
Comment (wlynch) [inline docs/gaps/6/README.md:123]: Do we want to wait for this?

**Review by wlynch — COMMENTED** [review]
Generally LGTM!

**Review by wlynch — APPROVED** [review]

File: docs/gaps/5/README.md (hunk 9 above)
Comment (adityasaky) [reply docs/gaps/5/README.md:115]: It's based on that, but if so, the RFC should be listed in the GAP process itself. I don't want us to put too fine a point on it.

File: docs/gaps/6/README.md (hunk 13 above)
Comment (adityasaky) [reply docs/gaps/6/README.md:123]: added!
